File: src/core/decoder.js
/**
 * decoder.js
 * Core decoding logic for received symbol grid.
 * Uses homography, modulation, packet, checksum.
 */

import { applyHomography } from './homography.js';
import { colorToValue, bitsPerSymbol } from './modulation.js';
import { HEADER_CELLS, gridGeom, cellCenter, parseHeader, valuesToBytes } from './packet.js';
import { verifyChecksum } from './checksum.js';

export class Decoder {
  constructor(density = 'medium') {
    this.density = density;
    this.geom = gridGeom(density);
    this.H = null; // homography
    this.lastFrameData = null;
  }

  setDensity(density) {
    this.density = density;
    this.geom = gridGeom(density);
  }

  setHomography(H) {
    this.H = H;
  }

  /**
   * Sample symbols from image data using homography.
   * imageData: ImageData or {data, width, height}
   * Returns array of symbol values or null on failure.
   */
  sampleSymbols(imageData, geom = this.geom) {
    if (!this.H) return null;

    const { width: W, height: Hh } = imageData;
    const data = imageData.data || imageData;

    const totalCells = geom.w * geom.h;
    const vals = [];
    let bad = false;

    for (let row = 0; row < geom.h && !bad; row++) {
      for (let col = 0; col < geom.w; col++) {
        const c = cellCenter(geom, col, row);
        const p = applyHomography(this.H, c.x, c.y);
        const px = Math.round(p.x);
        const py = Math.round(p.y);

        if (px < 0 || py < 0 || px >= W || py >= Hh) {
          bad = true;
          break;
        }

        const i = (py * W + px) * 4;
        const r = data[i], g = data[i + 1], b = data[i + 2];
        const v = colorToValue(r, g, b);
        vals.push(v);
      }
    }

    if (bad || vals.length < HEADER_CELLS) return null;
    return vals;
  }

  /**
   * Decode one frame's symbol values into {frameIdx, bytes, checksumValid, isFinal}
   */
  decodeFrame(symbolVals) {
    if (!symbolVals || symbolVals.length < HEADER_CELLS) return null;

    const { frameIdx, isFinal, checksum } = parseHeader(symbolVals);
    const dataVals = symbolVals.slice(HEADER_CELLS);
    const bytes = valuesToBytes(dataVals);
    const csValid = verifyChecksum(bytes, checksum);

    return {
      frameIdx,
      bytes,
      checksumValid: csValid,
      isFinal,
      rawChecksum: checksum
    };
  }

  /**
   * Full process frame: imageData -> decoded frame info
   */
  processFrame(imageData) {
    const vals = this.sampleSymbols(imageData);
    if (!vals) return null;

    const frameInfo = this.decodeFrame(vals);
    if (!frameInfo) return null;

    this.lastFrameData = { vals, frameInfo };
    return frameInfo;
  }
}

File: native/rust/src/decoder.rs
//! Core decoder logic (portable)

use crate::modulation::{ModulationMode, color_to_value};
use crate::packet::{GridGeom, parse_header, HEADER_CELLS};
use crate::homography::apply_homography;
use crate::checksum::verify_checksum;

#[derive(Clone, Debug)]
pub struct DecodedFrame {
    pub frame_idx: usize,
    pub bytes: Vec<u8>,
    pub checksum_valid: bool,
    pub is_final: bool,
}

pub struct Decoder {
    pub geom: GridGeom,
    pub homography: Option<[f32; 9]>,
    pub mode: ModulationMode,
}

impl Decoder {
    pub fn new(width: usize, height: usize, mode: ModulationMode) -> Self {
        let geom = GridGeom::new((width, height), 800.0, 600.0, 60.0);
        Self {
            geom,
            homography: None,
            mode,
        }
    }

    pub fn set_homography(&mut self, h: [f32; 9]) {
        self.homography = Some(h);
    }

    /// Sample symbols from raw RGB image data (width x height x 3 or x 4)
    pub fn sample_symbols(
        &self,
        image: &[u8],
        img_width: usize,
        img_height: usize,
        channels: usize,
    ) -> Option<Vec<u16>> {
        let h = self.homography?;
        let mut symbols = Vec::new();

        for row in 0..self.geom.height {
            for col in 0..self.geom.width {
                let (cx, cy) = self.geom.cell_center(col, row);
                let p = apply_homography(&h, cx, cy);

                let px = p.x.round() as isize;
                let py = p.y.round() as isize;

                if px < 0 || py < 0 || px as usize >= img_width || py as usize >= img_height {
                    return None;
                }

                let idx = ((py as usize) * img_width + px as usize) * channels;
                if idx + 2 >= image.len() {
                    return None;
                }

                let r = image[idx];
                let g = image[idx + 1];
                let b = image[idx + 2];

                let v = color_to_value(r, g, b, self.mode);
                symbols.push(v);
            }
        }
        Some(symbols)
    }

    pub fn decode_frame(&self, symbols: &[u16]) -> Option<DecodedFrame> {
        if symbols.len() < HEADER_CELLS {
            return None;
        }

        let (frame_idx, is_final, checksum) = parse_header(symbols)?;

        let data_vals = &symbols[HEADER_CELLS..];
        let mut bit_str = String::new();
        for &v in data_vals {
            bit_str.push_str(&format!("{:06b}", v)); // assumes 6-bit for now
        }

        let mut bytes = Vec::new();
        for i in (0..bit_str.len()).step_by(8) {
            if i + 8 <= bit_str.len() {
                if let Ok(b) = u8::from_str_radix(&bit_str[i..i + 8], 2) {
                    bytes.push(b);
                }
            }
        }

        let valid = verify_checksum(&bytes, checksum);

        Some(DecodedFrame {
            frame_idx,
            bytes,
            checksum_valid: valid,
            is_final,
        })
    }
}
